### PR TITLE
Feature/payments

### PR DESCRIPTION
**Pull request by paulshimmen**
This branch does the majority of the GovPay initiate payment code. Health check is working. Unit and integration tests are working. Payment is being inserted and updated into the database via the service project. A test index.html page has also been created which will test the re-direct to GovPay. The method to redirect is not ideal. I changed it for testing purposes, and also because I was getting CORs issues with swagger when testing.

### CODE REVIEW

**Review by manishagarwalEviden — CHANGES_REQUESTED** [review]

File: src/EPR.Payment.Facade/Controllers/Payments/PaymentsController.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            return BadRequest(ModelState);
+        }
+
+        try
+        {
+            var result = await _paymentsService.InitiatePaymentAsync(request, cancellationToken);
+
+            // Return a simple HTML page with JavaScript to perform the redirection
+            var htmlContent = $@"
+        <!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade/Controllers/Payments/PaymentsController.cs:45]: Ok, will create a private method as suggested.

File: src/EPR.Payment.Facade.Common/RESTServices/Payments/HttpGovPayService.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            if (_bearerToken != null)
+            {
+                SetBearerToken(_bearerToken); // Set the bearer token
+            }
+            else
+            {
+                throw new InvalidOperationException("Bearer token is null. Unable to initiate payment.");
+            }
+
+            var url = "payments";
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade.Common/RESTServices/Payments/HttpGovPayService.cs:37]: I will create a static constants class for the specific parts of the endpoint.

File: src/EPR.Payment.Facade.Common/RESTServices/Payments/HttpPaymentsService.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                config.Value.EndPointName ?? throw new ArgumentNullException(nameof(config), "PaymentService EndPointName configuration is missing"))
+        {
+            _httpContextAccessor = httpContextAccessor ?? throw new ArgumentNullException(nameof(httpContextAccessor));
+            _httpClientFactory = httpClientFactory ?? throw new ArgumentNullException(nameof(httpClientFactory));
+            _httpClientName = config.Value.HttpClientName ?? throw new ArgumentNullException(nameof(config), "PaymentService HttpClientName configuration is missing");
+        }
+
+        public async Task<Guid> InsertPaymentAsync(InsertPaymentRequestDto paymentStatusInsertRequest, CancellationToken cancellationToken)
+        {
+            var url = "payments";
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade.Common/RESTServices/Payments/HttpPaymentsService.cs:29]: I will create a static constants class for the specific parts of the endpoint.

File: src/EPR.Payment.Facade.Common/RESTServices/Payments/HttpPaymentsService.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            }
+            catch (Exception ex)
+            {
+                throw new Exception("Error occurred while inserting payment status.", ex);
+            }
+        }
+
+        public async Task UpdatePaymentAsync(Guid id, UpdatePaymentRequestDto paymentStatusUpdateRequest, CancellationToken cancellationToken)
+        {
+            var url = $"payments/{id}";
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade.Common/RESTServices/Payments/HttpPaymentsService.cs:43]: I will create a static constants class for the specific parts of the endpoint.

File: src/EPR.Payment.Facade.Common.UnitTests/RestServices/HttpGovPayServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                .Returns(new HttpClient(new Mock<HttpMessageHandler>().Object));
+
+            _httpGovPayService = new HttpGovPayService(
+                _httpContextAccessorMock.Object,
+                httpClientFactoryMock.Object,
+                _configMock.Object);
+        }
+
+        [TestMethod]
+        public async Task InitiatePayment_Success_ReturnsPaymentResponseDto()
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade.Common.UnitTests/RestServices/HttpGovPayServiceTests.cs]: I have added automoqdata for all tests where appropriate

File: src/EPR.Payment.Facade.Common.UnitTests/RestServices/HttpGovPayServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+                new HttpClientFactoryMock(httpClient),
+                _configMock!.Object);
+
+            var cancellationToken = new CancellationToken();
+
+            // Act
+            var result = await httpGovPayService.InitiatePaymentAsync(paymentRequestDto, cancellationToken);
+
+            // Assert
+            result.Should().NotBeNull();
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade.Common.UnitTests/RestServices/HttpGovPayServiceTests.cs]: I have reviewed all tests and added assertionscope  where appropriate.

File: src/EPR.Payment.Facade.Common.UnitTests/RestServices/HttpGovPayServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            result.Description.Should().Be(_expectedResponse.Description);
+            result.ReturnUrl.Should().Be(_expectedResponse.ReturnUrl);
+        }
+
+        [TestMethod]
+        public async Task InitiatePayment_Failure_ThrowsException()
+        {
+            // Arrange
+            var paymentRequestDto = new GovPayPaymentRequestDto
+            {
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade.Common.UnitTests/RestServices/HttpGovPayServiceTests.cs]: Have added automoqdata here

File: src/EPR.Payment.Facade.Common/Dtos/Request/Payments/GovPayPaymentRequestDto.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
+﻿using EPR.Payment.Facade.Common.Validators;
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade.Common/Dtos/Request/Payments/GovPayPaymentRequestDto.cs]: Renamed to GovPayRequestDto

File: src/EPR.Payment.Facade.Common.UnitTests/RestServices/HttpGovPayServiceTests.cs [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+                _configMock.Object);
+        }
+
+        [TestMethod]
+        public async Task InitiatePayment_Success_ReturnsPaymentResponseDto()
+        {
+            // Arrange
+            var paymentRequestDto = new GovPayPaymentRequestDto
+            {
+                Amount = 14500,
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade.Common.UnitTests/RestServices/HttpGovPayServiceTests.cs]: Done

File: src/EPR.Payment.Facade/Controllers/Payments/PaymentsController.cs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    [HttpPost]
+    [ProducesResponseType(StatusCodes.Status302Found)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+    [SwaggerOperation(Summary = "Initiates a new payment", Description = "Initiates a new payment with mandatory payment request data.")]
+    [SwaggerResponse(StatusCodes.Status302Found, "Redirects to the payment next URL.")]
+    [SwaggerResponse(StatusCodes.Status400BadRequest, "If the request is invalid or a validation error occurs.", typeof(ProblemDetails))]
+    [SwaggerResponse(StatusCodes.Status500InternalServerError, "If an unexpected error occurs.", typeof(ProblemDetails))]
+    [FeatureGate("EnablePaymentInitiation")]
+    public async Task<IActionResult> InitiatePayment([FromBody] PaymentRequestDto request, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (paulshimmen) [reply src/EPR.Payment.Facade/Controllers/Payments/PaymentsController.cs:33]: As discussed will leave this as dto object as it is actually a data transfer object.

**Review by manishagarwalEviden — APPROVED** [review]

**Pull request merged**